File: haskell/main_v1.hs
isMunchausen :: Int -> Int -> Int -> [Int] -> Bool
isMunchausen number n total cache
  | total > number = False
  | n > 0 = isMunchausen number (n `div` 10) (total + (cache !! (n `mod` 10))) cache
  | otherwise = number == total

main = do
  let cache = 0 : [i ^ i | i <- [1 .. 9]]
  mapM_ print $ filter (\i -> isMunchausen i i 0 cache) [0 .. 440000000]
  return 0

File: pascal/main_v1.pas
{$mode objfpc}{$H+}{$J-}

program Munchausen;

uses math;

const
  MAX = 440000000;


procedure CalcCache(var cache: array of LongInt);
var
  i: Integer;
begin
  cache[0] := 0;
  for i := 1 to 9 do
    cache[i] := i**i;
end;

function IsMunchausen(constref cache: array of LongInt; constref num: LongInt): Boolean;
var
  n, total, digit: LongInt;
begin
  n := num;
  total := 0;
  while n > 0 do
    begin
      DivMod(n, 10, n, digit);
      total += cache[digit];
      if total > num then
        break;
    end;
  Result := num = total;
end;

var
  cache: array[0..9] of LongInt;
  i: LongInt;
begin
  CalcCache(cache);
  for i := 0 to MAX-1 do
    if IsMunchausen(cache, i) then
      WriteLn(i);
end.
